community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/NativeSchemaNumberIndexReader.java

```diff
@@ -0,0 +1,165 @@
+import org.neo4j.kernel.api.schema.IndexQuery.ExactPredicate;
+import org.neo4j.kernel.api.schema.IndexQuery.NumberRangePredicate;
+import org.neo4j.kernel.configuration.Config;
+import org.neo4j.kernel.impl.api.index.sampling.IndexSamplingConfig;
+import org.neo4j.storageengine.api.schema.IndexReader;
+import org.neo4j.storageengine.api.schema.IndexSampler;
+
+class NativeSchemaNumberIndexReader<KEY extends NumberKey, VALUE extends NumberValue>
+        implements IndexReader
+{
+    private final GBPTree<KEY,VALUE> tree;
+    private final Layout<KEY,VALUE> layout;
+    private final KEY treeKeyFrom;
+    private final KEY treeKeyTo;
+    private RawCursor<Hit<KEY,VALUE>,IOException> openSeeker;
+
+    NativeSchemaNumberIndexReader( GBPTree<KEY,VALUE> tree, Layout<KEY,VALUE> layout )
+    {
+        this.tree = tree;
+        this.layout = layout;
+        this.treeKeyFrom = layout.newKey();
+        this.treeKeyTo = layout.newKey();
+    }
+
+    @Override
+    public void close()
+    {
+        ensureOpenSeekerClosed();
+    }
+
+    @Override
+    public IndexSampler createSampler()
+    {
+        // For an unique index there's an optimization, knowing that all values in it are unique, to simply count
+        // the number of indexes values and create a sample for that count. The GBPTree doesn't have an O(1)
+        // count mechanism, it will have to manually count the indexed values in it to get it.
+        // For that reason this implementation opts for keeping complexity down by just using the existing
+        // non-unique sampler which scans the index and counts (potentially duplicates, of which there will
+        // be none in a unique index).
+
+        IndexSamplingConfig indexSamplingConfig = new IndexSamplingConfig( Config.empty() );
+        FullScanNonUniqueIndexSampler<KEY,VALUE> sampler =
+                new FullScanNonUniqueIndexSampler<>( tree, layout, indexSamplingConfig );
+        return sampler::result;
+    }
+
+    @Override
+    public long countIndexedNodes( long nodeId, Object... propertyValues )
+    {
+        treeKeyFrom.from( nodeId, propertyValues );
+        treeKeyTo.from( nodeId, propertyValues );
+        try ( RawCursor<Hit<KEY,VALUE>,IOException> seeker = tree.seek( treeKeyFrom, treeKeyTo ) )
+        {
+            long count = 0;
+            while ( seeker.next() )
+            {
+                if ( seeker.get().key().entityId == nodeId )
+                {
+                    count++;
+                }
+            }
+            return count;
+        }
+        catch ( IOException e )
+        {
+            throw new UncheckedIOException( e );
+        }
+    }
+
+    @Override
+    public PrimitiveLongIterator query( IndexQuery... predicates ) throws IndexNotApplicableKernelException
+    {
+        if ( predicates.length != 1 )
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        ensureOpenSeekerClosed();
```

Comment:
Because we initially thought we only needed to support one open reader at any time, like how label scan store used to work (addressed in #9489 ). Will remove this.